mesonbuild/mtest.py

```diff
@@ -1859,17 +1860,34 @@ def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestS
         run all tests with that name across all subprojects, which is
         identical to "meson test foo1"
         '''
+        patterns = {}
         for arg in self.options.args:
             if ':' in arg:
                 subproj, name = arg.split(':', maxsplit=1)
+                if name == '':
+                    name = '*'
+                if subproj == '':  # in case arg was ':'
+                    subproj = '*'
             else:
-                subproj, name = '', arg
-            for t in tests:
-                if subproj and t.project_name != subproj:
-                    continue
-                if name and t.name != name:
-                    continue
-                yield t
+                subproj, name = '*', arg
+            patterns[(subproj, name)] = False
+
+        for t in tests:
+            for subprojp, namep in list(patterns):
+                if fnmatch(t.project_name, subprojp) and fnmatch(t.name, namep):
+                    patterns[(subprojp, namep)] = True
+                    yield t
+                    break
+
+        for (subprojp, namep), was_used in patterns.items():
+            if not was_used:
+                arg = f'{subproj}:{namep}'
```

Comment:
This seems broken, subproj is always the last iterated one from `for arg in self.options.args:`

Reply:
Nice catch, missing a `p` in the name indeed.
```suggestion
                arg = f'{subprojp}:{namep}'
```